src/tss2-fapi/ifapi_json_deserialize.c

```diff
@@ -969,19 +1075,180 @@ ifapi_json_IFAPI_EVENT_deserialize(json_object *jso,  IFAPI_EVENT *out)
+
+    if (verify && error_handling != DO_NOT_CHECK_DIGEST) {
+        r = check_event_digest(out, error_handling);
+        goto_if_error(r, "Check event digest", cleanup);
+    }
+    return TSS2_RC_SUCCESS;
+
+ cleanup:
+    if (out->content_type == IFAPI_IMA_EVENT_TAG) {
+        SAFE_FREE(out->content.ima_event.template_value.buffer);
+    } else if (out->content_type == IFAPI_PC_CLIENT) {
+         SAFE_FREE(out->content.firmware_event.data.buffer);
+    }
+    return r;
+}
+
+static char *field_TPMS_CEL_VERSION_tab[] = {
+    "major",
+    "minor",
+    "$schema"
+};
+
+/** Deserialize a TPMS_CEL_VERSION variable.
+ *
+ * @param[in]  jso json object to be deserialized.
+ * @param[out] out the deserialized object.
+ * @retval TSS2_FAPI_RC_BAD_VALUE if an invalid value was passed into
+ *         the function.
+ */
+TSS2_RC
+ifapi_json_TPMS_CEL_VERSION_deserialize(json_object *jso,
+        TPMS_CEL_VERSION *out)
+{
+    LOG_TRACE("call");
+    json_object *jso2;
+    TSS2_RC r;
+
+    memset(out, 0, sizeof(TPMS_CEL_VERSION));
+    ifapi_check_json_object_fields(jso, &field_TPMS_CEL_VERSION_tab[0],
+                                   SIZE_OF_ARY(field_TPMS_CEL_VERSION_tab));
+    if (!ifapi_get_sub_object(jso, "major", &jso2)) {
+        LOG_ERROR("Field \"major\" not found.");
+        return TSS2_FAPI_RC_BAD_VALUE;
+    }
+    r = ifapi_json_UINT16_deserialize(jso2, &out->major);
+    return_if_error(r, "Bad value for field \"major\".");
+
+    if (!ifapi_get_sub_object(jso, "minor", &jso2)) {
+        LOG_ERROR("Field \"minor\" not found.");
+        return TSS2_FAPI_RC_BAD_VALUE;
+    }
+    return ifapi_json_UINT16_deserialize(jso2, &out->minor);
+}
+
+/** Deserialize a TPMU_CELMGT json object.
+ *
+ * Desearialize a canonical eventlog management event.
+ * @param[in]  selector The CEL type.
+ * @param[in]  jso the json object to be deserialized.
+ * @param[out] out the deserialzed binary object.
+ * @retval TSS2_RC_SUCCESS if the function call was a success.
+ * @retval TSS2_FAPI_RC_BAD_VALUE if the json object can't be deserialized.
+ * @retval TSS2_FAPI_RC_BAD_REFERENCE a invalid null pointer is passed.
+ */
+TSS2_RC
+ifapi_json_TPMU_CELMGT_deserialize(
+    UINT32 selector,
+    json_object *jso,
+    TPMU_CELMGT *out)
+{
+    json_object *jso2 = NULL;
+    LOG_TRACE("call");
+    switch (selector) {
+    case CEL_VERSION:
+        return ifapi_json_TPMS_CEL_VERSION_deserialize(jso, &out->cel_version);
+    case FIRMWARE_END:
+        return ifapi_json_TPMS_EMPTY_deserialize(jso, &out->firmware_end);
+    case CEL_TIMESTAMP:
+        if (!ifapi_get_sub_object(jso, "cel_timestamp", &jso2)) {
+            LOG_ERROR("Field \"cel_timestamp\" not found.");
+            return TSS2_FAPI_RC_BAD_VALUE;
+        }
+        return ifapi_json_UINT64_deserialize(jso2, &out->cel_timestamp);
+
+    default:
+        LOG_TRACE("false");
+        return TSS2_FAPI_RC_BAD_VALUE;
+    };
+}
+
```

Comment:
extra newline or was that intentional as a visual break between functions for the static array data?